HWMv2: Port iMX M series cores to HWMv2

danieldegrasse opened the pull request:
This PR ports all iMX M series cores to HWMv2. It also removes the legacy `boards/boards_legacy/arm` folder, since there are no longer any ARM boards using HWMv1

*Note- I did not remove the `soc/soc_legacy/arm` folder*, as some of the Kconfigs defined in `soc/soc_legacy/arm/Kconfig` are still in use in tree. I'm unclear on where best to define these in HWMv2, but happy to add the change to the PR if there is a preferred location.

This PR also contains a commit to fix a build error on the RT11xx SOC series. I can split this into a separate PR if desired, but I included it in this PR given the tight time constraint for HWMv2

danieldegrasse committed:
soc: nxp: convert iMX6 SoloX core to HWMv2

Convert iMX6 SoloX core to HWMv2. This port currently only supports the
M4 core (as it did in HWMv1)

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: udoo_neo_full: convert to HWMv2

Convert udoo_neo_full board to HWMv2.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: nxp: convert iMX7 Dual core to HWMv2

Convert iMX7 Dual core to HWMv2. As in HWMv1, Only M4 core is supported
by this port.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: colibri_imx7d: port to HWMv2

Port Colibri iMX7D board to HWMv2. This port only supports the M4 core
of the iMX7D SOC.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: meerkat96: port to HWMv2

Port 96boards meerkat96b board to HWMv2. This board port only supports
the M4 core present on the iMX7D SOC.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: warp7: convert to HWMv2

Convert warp7 board to HWMv2. This board was originally ported
using the iMX7D SOC (despite the fact that board uses an iMX7S
SOC). For HWMv2, this board continues to use the IMX7D SOC port,
since only the M4 core is enabled.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: pico_pi: convert to HVMv2

Convert pico_pi board to HWMv2. This board port only supports the M4
core on the SOC

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: nxp: imx8mm: port iMX8MM M4 core to HVMv2

Port iMX8MM M4 core of iMX8MM SOC  to HVMv2. The A53 core has already
been ported, so this port adds to that SOC definition.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: imx8mm_evk: port CM4 core to HWMv2

Port CM4 core to HVMv2. This core is merged with the existing board
definition, which supported the A53 target.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimx8mm_phyboard: convert to HVMv2

Convert mimx8mm_phyboard to HVMv2. This port only enables the M4 core of
the iMX8MM SOC present on this board.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: nxp: imx8mp: port M7 core to HWMv2

Port M7 core of imx8mp to HWMv2

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: nxp: mimx8mp_evk: port M7 core to HWMv2

Port M7 core of mimx8mp_evk to HWMv2

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: phyboard_pollux: move to HVMv2

Move phyboard_pollux to HWMv2

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: verdin_imx8mp: convert to HVMv2

Convert verdin_imx8mp to HWMv2. Only the M7 core of the iMX8MP is
currently supported

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: nxp: port imx8mq SOC to HWMv2

Port IMX8M Quad SOC to HWMv2. Only the M4 core is enabled.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimx8mq_evk: port to HWMv2

Port mimx8mq_evk to HWMv2. As this is the final SOC in the legacy
boards folder, that folder is also removed

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: nxp: imxrt11xx: only set CONFIG_CPU_HAS_FPU_DOUBLE_PRECISION for M7

Only set CONFIG_CPU_HAS_FPU_DOUBLE_PRECISION for Cortex M7 cores, as the
M4 core on iMXRT11xx parts does have a double precision FPU

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

JiafeiPan commented on boards/nxp/imx8mp_evk/imx8mp_evk_mimx8ml8_m7_ddr.yaml:
```diff
@@ -4,7 +4,7 @@
 #
 
-identifier: mimx8mp_evk_ddr
+identifier: mimx8mp_evk/mimx8ml8/m7/ddr
```
okay, it has already been fixed.

danieldegrasse commented on soc/nxp/imx/soc.yml:
```diff
@@ -38,9 +38,29 @@ family:
+      - name: a53_cpu0
+      - name: a53_cpu1
+      - name: a53_cpu2
+      - name: a53_cpu3
```
This is typically how we define CPU clusters for dual core parts using the same core type on MCUs. Would you prefer that only one `a53` target be defined?

nordicjm replied:
SMP clusters just define a single `a53`

danieldegrasse commented on drivers/clock_control/clock_control_mcux_syscon.c:
```diff
@@ -27,7 +27,7 @@ static int mcux_lpc_syscon_clock_control_on(const struct device *dev,
 #if defined(CONFIG_SOC_FAMILY_LPC)
 		CLOCK_EnableClock(kCLOCK_Mrt);
-#elif defined(CONFIG_SOC_FAMILY_IMX)
+#elif defined(SOC_FAMILY_NXP_IMXRT)
```
I'm not sure I follow here- I'm using this Kconfig because (to my knowledge) the SYSCON block is only present on LPC and NXP iMX RT parts, so this config should be enough to determine when to use the given `CLOCK` API

pdgendt commented on drivers/clock_control/clock_control_mcux_syscon.c:
```diff
@@ -27,7 +27,7 @@ static int mcux_lpc_syscon_clock_control_on(const struct device *dev,
 #if defined(CONFIG_SOC_FAMILY_LPC)
 		CLOCK_EnableClock(kCLOCK_Mrt);
-#elif defined(CONFIG_SOC_FAMILY_IMX)
+#elif defined(SOC_FAMILY_NXP_IMXRT)
```
Isn't this missing a `CONFIG_` prefix?

nordicjm commented on tests/drivers/gpio/gpio_basic_api/src/main.c:
```diff
@@ -10,7 +10,7 @@
 #if defined(CONFIG_BOARD_RV32M1_VEGA)
 #include <fsl_port.h>
-#elif defined(CONFIG_BOARD_UDOO_NEO_FULL_M4)
+#elif defined(BOARD_UDOO_NEO_FULL_MCIMX6X_M4)
```
missing CONFIG_

nordicjm commented on tests/drivers/gpio/gpio_basic_api/src/main.c:
```diff
@@ -32,7 +32,7 @@ static void board_setup(void)
 #endif
 
-#if defined(CONFIG_BOARD_UDOO_NEO_FULL_M4)
+#if defined(BOARD_UDOO_NEO_FULL_MCIMX6X_M4)
```
as above

nordicjm commented on soc/nxp/imx/soc.yml:
```diff
@@ -50,3 +50,10 @@ family:
+    socs:
+    - name: mcimx7d
+      cpuclusters:
+      - name: a7_cpu0
+      - name: a7_cpu1
```
should only have a single `a7` as these are SMP

nordicjm commented on tests/drivers/pwm/pwm_api/src/test_pwm.c:
```diff
@@ -54,7 +54,7 @@
 #endif
 
-#if defined(CONFIG_BOARD_COLIBRI_IMX7D_M4) || defined(CONFIG_SOC_MK64F12) ||                       \
+#if defined(BOARD_COLIBRI_IMX7D_MCIMX7D_M4) || defined(CONFIG_SOC_MK64F12) ||                      \
```
as above

nordicjm commented on boards/96boards/meerkat96/doc/index.rst:
```diff
@@ -109,7 +109,7 @@ features:
 The default configuration can be found in the defconfig file:
 
-	``boards/arm/96b_meerkat96/96b_meerkat96_defconfig``
+	:zephyr_file:`boards/96boards/96b_meerkat96/96b_meerkat96_mcimx7d_m4_defconfig`
```
invalid

nordicjm commented on boards/96boards/meerkat96/doc/index.rst:
```diff
@@ -178,7 +178,7 @@ For more information about memory mapping see the
 
 At compilation time you have to choose which RAM will be used. This
-configuration is done in the file ``boards/arm/96b_meerkat96/96b_meerkat96.dts``
+configuration is done in the file :zephyr_file:`boards/96boards/96b_meerkat96/96b_meerkat96_mcimx7d_m4.dts`
```
invalid

nordicjm commented on boards/phytec/mimx8mm_phyboard_polis/doc/index.rst:
```diff
@@ -1,4 +1,4 @@
-.. _mimx8mm_phyboard_polis:
+.. _mimx8mm_phyboard_polis/mimx8mm6/m4:
```
??

danieldegrasse replied:
Caught in a find and replace while renaming cores

nordicjm commented on boards/phytec/mimx8mm_phyboard_polis/doc/index.rst:
```diff
@@ -70,7 +70,7 @@ More information about the board can be found at the
 ==================
 
-The Zephyr mimx8mm_phyboard_polis board configuration supports the following
+The Zephyr mimx8mm_phyboard_polis/mimx8mm6/m4 board configuration supports the following
```
enclose name in double `

pdgendt reviewed (approved):
Comments before rebase appear to be fixed

nordicjm commented:
Compliance failure

danieldegrasse commented:
> Compliance failure

See comment above- I think this is because no legacy boards exist anymore in this PR

nordicjm commented:
`soc/Kconfig.v1` probably needs the bottom line removing and `scripts/ci/check_compliance.py` probably needs this removing:
```
                fp.write('source "boards/boards_legacy/*/*/Kconfig.defconfig"\n')
                fp.write('osource "soc/soc_legacy/*/*/Kconfig.defconfig"\n')
```

nordicjm commented:
Actually remove this:
```
        if hwm is None or hwm == "v1":
            v1_file = os.path.join(kconfiglib_dir, "Kconfig.v1")
            v1_syms_file = os.path.join(kconfiglib_boards_dir, 'Kconfig.syms.v1')
            with open(v1_file, 'w') as fp:
                fp.write('source "boards/boards_legacy/*/*/Kconfig.defconfig"\n')
                fp.write('osource "soc/soc_legacy/*/*/Kconfig.defconfig"\n')
                fp.write('source "boards/Kconfig"\n')
                fp.write('source "soc/Kconfig"\n')

            os.environ["BOARD_DIR"] = "boards/boards_legacy/*/*"
            self.get_v1_model_syms(v1_file, v1_syms_file)

        if hwm is None or hwm == "v2":
```

And left indent these 2 lines:
```
            os.environ["BOARD_DIR"] = kconfiglib_boards_dir
            self.get_v2_model(kconfiglib_dir)
```

danieldegrasse committed:
scripts: ci: check_compliance: remove HWMv1 checks

Remove HWMv1 check compliance code, since no HWMv1 boards or SOCs
now exist in tree

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>
Co-authored-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

danieldegrasse commented:
> Actually remove this:

Done- added a commit to this PR, and made you a co-author, hope that's fine

danieldegrasse commented:
> Note- I did not remove the soc/soc_legacy/arm folder, as some of the Kconfigs defined in soc/soc_legacy/arm/Kconfig are still in use in tree. I'm unclear on where best to define these in HWMv2, but happy to add the change to the PR if there is a preferred location.

@nordicjm any thoughts on this? We can also leave this to a follow up PR
